node/libs/storage/src/block_store.rs

```diff
@@ -0,0 +1,218 @@
+    /// Consensus code calls this method only once and then tracks the
+    /// range of avaliable blocks internally.
+    async fn state(&self, ctx: &ctx::Ctx) -> ctx::Result<Option<BlockStoreState>>;
+
+    /// Gets a block by its number. Returns None if block is missing.
+    async fn block(
+        &self,
+        ctx: &ctx::Ctx,
+        number: validator::BlockNumber,
+    ) -> ctx::Result<Option<validator::FinalBlock>>;
+
+    /// Persistently store a block.
+    /// Implementations are only required to accept a block directly after the current last block,
+    /// so that the stored blocks always constitute a continuous range.
+    /// Implementation should return only after the block is stored PERSISTENTLY -
+    /// consensus liveness property depends on this behavior.
+    async fn store_next_block(
+        &self,
+        ctx: &ctx::Ctx,
+        block: &validator::FinalBlock,
+    ) -> ctx::Result<()>;
+}
+
+#[derive(Debug)]
+struct Inner {
+    inmem: sync::watch::Sender<BlockStoreState>,
+    persisted: BlockStoreState,
+    // TODO: this data structure can be optimized to a VecDeque (or even just a cyclical buffer).
+    cache: BTreeMap<validator::BlockNumber, validator::FinalBlock>,
+    cache_capacity: usize,
+}
+
+/// A wrapper around a PersistentBlockStore which adds caching blocks in-memory
+/// and other useful utilities.
+#[derive(Debug)]
+pub struct BlockStore {
+    inner: sync::watch::Sender<Inner>,
+    persistent: Box<dyn PersistentBlockStore>,
+}
+
+/// Runner of the BlockStore background tasks.
+#[must_use]
+pub struct BlockStoreRunner(Arc<BlockStore>);
+
+impl BlockStoreRunner {
+    /// Runs the background tasks of the BlockStore.
+    pub async fn run(self, ctx: &ctx::Ctx) -> anyhow::Result<()> {
+        let res = async {
+            let inner = &mut self.0.inner.subscribe();
+            loop {
+                let block = sync::wait_for(ctx, inner, |inner| !inner.cache.is_empty())
+                    .await?
+                    .cache
+                    .first_key_value()
+                    .unwrap()
+                    .1
+                    .clone();
+                self.0.persistent.store_next_block(ctx, &block).await?;
+                self.0.inner.send_modify(|inner| {
+                    debug_assert!(inner.persisted.next() == block.header().number);
+                    inner.persisted.last = block.justification.clone();
+                    inner.cache.remove(&block.header().number);
+                });
+            }
+        }
+        .await;
+        match res {
+            Ok(()) | Err(ctx::Error::Canceled(_)) => Ok(()),
+            Err(ctx::Error::Internal(err)) => Err(err),
+        }
+    }
+}
+
+impl BlockStore {
+    /// Constructs a BlockStore.
+    /// BlockStore takes ownership of the passed PersistentBlockStore,
+    /// i.e. caller should modify the underlying persistent storage
+    /// (add/remove blocks) ONLY through the constructed BlockStore.
```

Comment:
Dumb question: Is removing blocks planned in the future? AFAIU, it's not implemented right now.

Reply:
we will most likely gc blocks at some point in the future. Then we will need the consensus to be aware that old blocks are not available any more. How it will be implemented is tbd. I've dropped the "add/remove" from the comment for now.